exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceInstanceSpec.java

```diff
@@ -0,0 +1,35 @@
+package com.exonum.binding.core.runtime;
+
+import com.exonum.binding.common.message.TransactionMessage;
+import com.google.auto.value.AutoValue;
+
+/**
+ * A specification of a service instance.
+ */
+@AutoValue
+public abstract class ServiceInstanceSpec {
+
+  /**
+   * Returns the name of the service instance. It serves as the primary identifier of this service
+   * in most operations. It is assigned by the network administrators.
+   */
+  public abstract String getName();
+
+  /**
+   * Returns the numeric id of the service instance. Exonum assigns it to the service
```

Comment:
> Exonum assigns it to the service on instantiation.

As I remember, a user must provide id for its service which is unique within the network. Does it change?

Reply:
Yes, it has been changed in dynamic services. As administrators can create several instances from the same artifact, the service id cannot be hardcoded in the service definition, but must be assigned on instantiation, so that the transaction messages can be routed to the correct instance. The id assignment is done by the core; the name is provided by the administrators (and is considered the primary identifier of the service instance).